src/iop/rawprepare.c

```diff
@@ -0,0 +1,473 @@
+               const dt_iop_roi_t *const roi_in, const dt_iop_roi_t *const roi_out)
+{
+  dt_iop_rawprepare_data_t *d = (dt_iop_rawprepare_data_t *)piece->data;
+  dt_iop_rawprepare_global_data_t *gd = (dt_iop_rawprepare_global_data_t *)self->data;
+
+  const int devid = piece->pipe->devid;
+  cl_mem dev_sub = NULL;
+  cl_mem dev_div = NULL;
+  cl_int err = -999;
+
+  int kernel = -1;
+
+  if(!dt_dev_pixelpipe_uses_downsampled_input(piece->pipe) && dt_image_filter(&piece->pipe->image))
+  {
+    kernel = gd->kernel_rawprepare_1f;
+  }
+  else
+  {
+    kernel = gd->kernel_rawprepare_4f;
+  }
+
+  dev_sub = dt_opencl_copy_host_to_device_constant(devid, sizeof(float) * 4, d->sub);
+  if(dev_sub == NULL) goto error;
+
+  dev_div = dt_opencl_copy_host_to_device_constant(devid, sizeof(float) * 4, d->div);
+  if(dev_div == NULL) goto error;
+
+  const int width = roi_out->width;
+  const int height = roi_out->height;
+
+  size_t sizes[] = { ROUNDUPWD(roi_in->width), ROUNDUPHT(roi_in->height), 1 };
+  dt_opencl_set_kernel_arg(devid, kernel, 0, sizeof(cl_mem), (void *)&dev_in);
+  dt_opencl_set_kernel_arg(devid, kernel, 1, sizeof(cl_mem), (void *)&dev_out);
+  dt_opencl_set_kernel_arg(devid, kernel, 2, sizeof(int), (void *)&(width));
+  dt_opencl_set_kernel_arg(devid, kernel, 3, sizeof(int), (void *)&(height));
+  dt_opencl_set_kernel_arg(devid, kernel, 4, sizeof(cl_mem), (void *)&dev_sub);
+  dt_opencl_set_kernel_arg(devid, kernel, 5, sizeof(cl_mem), (void *)&dev_div);
+  dt_opencl_set_kernel_arg(devid, kernel, 6, sizeof(uint32_t), (void *)&roi_out->x);
+  dt_opencl_set_kernel_arg(devid, kernel, 7, sizeof(uint32_t), (void *)&roi_out->y);
+  err = dt_opencl_enqueue_kernel_2d(devid, kernel, sizes);
+  if(err != CL_SUCCESS) goto error;
+
+  dt_opencl_release_mem_object(dev_sub);
+  dt_opencl_release_mem_object(dev_div);
+
+  return TRUE;
+
+error:
+  if(dev_sub != NULL) dt_opencl_release_mem_object(dev_sub);
+  if(dev_div != NULL) dt_opencl_release_mem_object(dev_div);
+  dt_print(DT_DEBUG_OPENCL, "[opencl_rawprepare] couldn't enqueue kernel! %d\n", err);
+  return FALSE;
+}
+#endif
+
+void commit_params(dt_iop_module_t *self, const dt_iop_params_t *const params, dt_dev_pixelpipe_t *pipe,
+                   dt_dev_pixelpipe_iop_t *piece)
+{
+  const dt_iop_rawprepare_params_t *const p = (dt_iop_rawprepare_params_t *)params;
+  dt_iop_rawprepare_data_t *d = (dt_iop_rawprepare_data_t *)piece->data;
+
+  if(!dt_dev_pixelpipe_uses_downsampled_input(piece->pipe) && dt_image_filter(&piece->pipe->image))
+  {
+    const float white = (float)p->raw_white_point;
+
+    for(int i = 0; i < 4; i++)
+    {
+      d->sub[i] = (float)p->raw_black_level_separate[i];
+      d->div[i] = (white - d->sub[i]);
+    }
+  }
+  else
+  {
+    const float white = (float)p->raw_white_point / (float)UINT16_MAX;
+    float black = 0;
+    for(int i = 0; i < 4; i++)
+    {
+      black += p->raw_black_level_separate[i] / (float)UINT16_MAX;
+    }
+    black /= 4.0f;
```

Comment:
This seems wrong. black_level_separate means that different channels have different black levels, yet we are just averaging them.